consensus/hotstuff/signature/weighted_signature_aggregator_test.go

```diff
@@ -0,0 +1,261 @@
+		}
+		signers, agg, err = aggregator.Aggregate()
+		assert.NoError(t, err)
+		ok, err = verifyAggregate(signers, ids, agg, msg, tag)
+		assert.NoError(t, err)
+		assert.True(t, ok)
+		// check signers
+		signers = sortIdentifiers(signers)
+		for i := 0; i < signersNum; i++ {
+			assert.Equal(t, signers[i], ids[i].NodeID)
+		}
+	})
+
+	invalidInput := engine.NewInvalidInputError("some error")
+	duplicate := engine.NewDuplicatedEntryErrorf("some error")
+	invalidSig := signature.ErrInvalidFormat
+
+	// Unhappy paths
+	t.Run("invalid signer ID", func(t *testing.T) {
+		aggregator, _, sigs, _, _ := createAggregationData(t, signersNum)
+		// generate an ID that is not in the node ID list
+		invalidId := unittest.IdentityFixture().NodeID
+
+		err := aggregator.Verify(invalidId, sigs[0])
+		assert.Error(t, err)
+		assert.IsType(t, invalidInput, err)
+
+		weight, err := aggregator.TrustedAdd(invalidId, sigs[0])
+		assert.Equal(t, uint64(0), weight)
+		assert.Equal(t, uint64(0), aggregator.TotalWeight())
+		assert.Error(t, err)
+		assert.IsType(t, invalidInput, err)
+	})
+
+	t.Run("duplicate signature", func(t *testing.T) {
+		aggregator, ids, sigs, _, _ := createAggregationData(t, signersNum)
+		expectedWeight := uint64(0)
+		// add signatures
+		for i, sig := range sigs {
+			weight, err := aggregator.TrustedAdd(ids[i].NodeID, sig)
+			expectedWeight += ids[i].Stake
+			assert.Equal(t, expectedWeight, weight)
+			require.NoError(t, err)
+		}
+		// add same duplicates and test thread safety
+		var wg sync.WaitGroup
+		for i, sig := range sigs {
+			wg.Add(1)
+			// test thread safety
+			go func(i int, sig crypto.Signature) {
+				defer wg.Done()
+				weight, err := aggregator.TrustedAdd(ids[i].NodeID, sigs[i]) // same signature for same index
+				// weight should not change
+				assert.Equal(t, expectedWeight, weight)
+				assert.Error(t, err)
+				assert.IsType(t, duplicate, err)
+				weight, err = aggregator.TrustedAdd(ids[i].NodeID, sigs[(i+1)%signersNum]) // different signature for same index
+				// weight should not change
+				assert.Equal(t, expectedWeight, weight)
+				assert.Error(t, err)
+				assert.IsType(t, duplicate, err)
+			}(i, sig)
+		}
+		wg.Wait()
+	})
+
+	t.Run("invalid signature", func(t *testing.T) {
+		aggregator, ids, sigs, _, _ := createAggregationData(t, signersNum)
+		// corrupt sigs[0]
+		sigs[0][4] ^= 1
+		// test Verify
+		err := aggregator.Verify(ids[0].NodeID, sigs[0])
+		require.Error(t, err)
+		assert.True(t, errors.Is(err, invalidSig))
+
+		// add signatures for aggregation including corrupt sigs[0]
+		expectedWeight := uint64(0)
+		for i, sig := range sigs {
+			weight, err := aggregator.TrustedAdd(ids[i].NodeID, sig)
+			require.NoError(t, err)
+			expectedWeight += ids[i].Stake
+			assert.Equal(t, expectedWeight, weight)
+		}
+		signers, agg, err := aggregator.Aggregate()
+		assert.Error(t, err)
+		assert.Nil(t, agg)
+		assert.Nil(t, signers)
+		// fix sigs[0]
+		sigs[0][4] ^= 1
```

Comment:
why we need this?

Reply:
It alters one bit of the signature to make it invalid. It also tests a specific error case because the resulting signature is not a valid curve point